src/wallet/wallet.cpp

```diff
@@ -3219,21 +3217,16 @@ void CWallet::UpdateSaplingNullifierNoteMapWithTx(CWalletTx& wtx) {
         else {
             uint64_t position = nd.witnesses.front().position();
             auto extfvk = mapSaplingFullViewingKeys.at(nd.ivk);
-            OutputDescription output = wtx.vShieldedOutput[op.n];
 
-            auto optDeserialized = SaplingNotePlaintext::attempt_sapling_enc_decryption_deserialization(output.encCiphertext, nd.ivk, output.ephemeralKey);
+            auto optDecrypted = wtx.DecryptSaplingNote(Params(), op);
 
             // The transaction would not have entered the wallet unless
             // its plaintext had been successfully decrypted previously.
-            assert(optDeserialized != std::nullopt);
-
-            auto optPlaintext = SaplingNotePlaintext::plaintext_checks_without_height(*optDeserialized, nd.ivk, output.ephemeralKey, output.cmu);
+            assert(optDecrypted != std::nullopt);
+            SaplingNotePlaintext notePt;
+            std::tie(notePt, std::ignore) = optDecrypted.value();
 
-            // An item in mapSaplingNoteData must have already been successfully decrypted,
-            // otherwise the item would not exist in the first place.
-            assert(optPlaintext != std::nullopt);
-
-            auto optNote = optPlaintext.value().note(nd.ivk);
+            auto optNote = notePt.note(nd.ivk);
```

Comment:
[SaplingNotePlaintext::note(const SaplingIncomingViewingKey& ivk)](https://github.com/zcash/zcash/blob/a37c1b06ec1db0ee00b50a57bbb3dd5317dab9d8/src/zcash/Note.cpp#L175-L189) is recalculating the address (for `pk_d`), which it should not have to do because we just calculated it.

Reply:
Please file a ticket so we don't lose track.